test/end-to-end/oauth2provider.test.js

```diff
@@ -0,0 +1,187 @@
+ * obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Imports
+ */
+
+import assert from "assert";
+import puppeteer from "puppeteer";
+import {
+    clearBrowserCookiesWithoutAffectingConsole,
+    toggleSignInSignUp,
+    screenshotOnFailure,
+    backendBeforeEach,
+    waitForUrl,
+    createOAuth2Client,
+    setOAuth2ClientIdInStorage,
+    removeOAuth2ClientIdFromStorage,
+    getOAuth2LoginButton,
+    getOAuth2LogoutButton,
+    getOAuth2TokenData,
+    isReact16,
+    waitFor,
+    signUp,
+    getDefaultSignUpFieldValues,
+    getTestEmail,
+} from "../helpers";
+import fetch from "isomorphic-fetch";
+
+import { TEST_CLIENT_BASE_URL, TEST_SERVER_BASE_URL, SIGN_OUT_API } from "../constants";
+
+/*
+ * Tests.
+ */
+describe("SuperTokens OAuth2Provider", function () {
+    let browser;
+    let page;
+    let consoleLogs = [];
+
+    before(async function () {
+        // Skip these tests if running in React 16
+        if (isReact16()) {
+            this.skip();
+        }
+
+        await backendBeforeEach();
+
+        await fetch(`${TEST_SERVER_BASE_URL}/startst`, {
+            method: "POST",
+        }).catch(console.error);
+
+        browser = await puppeteer.launch({
+            args: ["--no-sandbox", "--disable-setuid-sandbox"],
+            headless: true,
+        });
+    });
+
+    after(async function () {
+        await browser.close();
+
+        await fetch(`${TEST_SERVER_BASE_URL}/after`, {
+            method: "POST",
+        }).catch(console.error);
+
+        await fetch(`${TEST_SERVER_BASE_URL}/stopst`, {
+            method: "POST",
+        }).catch(console.error);
+    });
+
+    afterEach(function () {
+        return screenshotOnFailure(this, browser);
+    });
+
+    beforeEach(async function () {
+        page = await browser.newPage();
+        page.on("console", (consoleObj) => {
+            const log = consoleObj.text();
+            if (log.startsWith("ST_LOGS")) {
+                consoleLogs.push(log);
+            }
+        });
+        consoleLogs = await clearBrowserCookiesWithoutAffectingConsole(page, []);
+    });
+
+    describe("Generic OAuth2 Client Library", function () {
+        afterEach(async function () {
+            await removeOAuth2ClientIdFromStorage(page);
+        });
+
+        it("should successfully complete the OAuth2 flow", async function () {
+            const { client } = await createOAuth2Client({
+                scope: "offline_access profile openid email",
+                redirectUris: [`${TEST_CLIENT_BASE_URL}/oauth2/callback`],
```

Comment:
Please add a note somewhere in this file about using the same URL (and why we do not need to have a separate URL for the flow to work right)

Reply:
Not sure what you meant. We use different url but the component/page is the same. I have left a comment about that in the OAuth2Page component.

Reply:
Ah right, I meant why we are using the same domain.